### PR TITLE
Add frobenius_norm method to BlockMatrixBase

### PR DESCRIPTION
**Pull request by class4kayaker**
Fix to #7629

I could not find any tests in `tests/lac/` which used the `BlockMatrixBase::matrix_norm_square` function, so while I would like to add a test for this new functionality, I do not see an obvious place to insert it.

### CODE REVIEW

**Comment by bangerth** [comment]
Do you think you could add an entry to `doc/news/changes/minor` that explains your change?

**Comment by masterleinad** [comment]
> I could not find any tests in `tests/lac/` which used the `BlockMatrixBase::matrix_norm_square` function, so while I would like to add a test for this new functionality, I do not see an obvious place to insert it.

`lac/block_matrices_05.cc` would probably be suitable.

**Comment by class4kayaker** [comment]
Test added, it may also be a suitable location to insert other basic block matrix tests.

File: tests/lac/block_matrices_05.cc [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+
+
+int
+main()
+{
+  std::ofstream logfile("output");
+  deallog << std::fixed;
+  deallog << std::setprecision(2);
+  deallog.attach(logfile);
```
Comment (masterleinad) [inline tests/lac/block_matrices_05.cc]: Would you mind replacing the four lines above by `initlog()`?

File: tests/lac/block_matrices_05.cc [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  const double accuracy          = 1e-12;
+  const double correct_frob_norm = 10.0;
+  const double frob_norm         = bsm.frobenius_norm();
+
+  Assert(std::fabs(frob_norm - correct_frob_norm) < accuracy,
+         ExcInternalError());
+
+  deallog << "OK" << std::endl;
+  deallog << std::flush;
```
Comment (masterleinad) [inline tests/lac/block_matrices_05.cc]:
```suggestion
```
No need to use `std::flush` here. `std::endl` flushes the buffer already.

File: include/deal.II/lac/block_matrix_base.h [hunk 3]
```diff
@@ -2401,6 +2408,29 @@ BlockMatrixBase<MatrixType>::matrix_norm_square(const BlockVectorType &v) const
 
 
 
+template <class MatrixType>
+typename BlockMatrixBase<MatrixType>::value_type
+BlockMatrixBase<MatrixType>::frobenius_norm() const
+{
+  value_type norm_sqr = 0;
+
+
```
Comment (masterleinad) [inline include/deal.II/lac/block_matrix_base.h]:
```suggestion
```

File: tests/lac/block_matrices_05.cc [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Test that the frobenius_norm function returns the correct value
+// for a block matrix
```
Comment (masterleinad) [inline tests/lac/block_matrices_05.cc]:
```suggestion
// for a block matrix.
```

File: doc/news/changes/minor/20190806JonathanRobey [hunk 5]
```diff
@@ -0,0 +1,2 @@
+New: Added frobenius_norm method to BlockMatrixBase
```
Comment (masterleinad) [inline doc/news/changes/minor/20190806JonathanRobey]:
```suggestion
New: Added frobenius_norm method to BlockMatrixBase.
```

**Comment by bangerth** [comment]
On 8/6/19 4:57 PM, Jonathan Robey wrote:
> Test added, it may also be a suitable location to insert other basic block 
> matrix tests.

Follow-up patches are always welcome! :-)

**Comment by class4kayaker** [comment]
Comments addressed, I can rebase and squash if desired.

**Comment by drwells** [comment]
I have the same comment here as in #8487: we should use a real type for `frobenius_norm`, like we do for `Vector<T>` and `FullMatrix<T>`.

**Comment by class4kayaker** [comment]
Changed to use a real_type.

**Comment by drwells** [comment]
Looks good to me.

We need to be careful since we added `real_type` to both block matrix PRs; lets merge this one first and then handle the other one to avoid a conflict (and extra test suite run).

/rebuild

**Comment by class4kayaker** [comment]
Agreed on the order and likely merge conflict.

**Comment by drwells** [comment]
@masterleinad are you happy with these changes? If so, lets merge.

**Pull request merged**